aspnetcore/host-and-deploy/health-checks.md

```diff
@@ -0,0 +1,359 @@
+For many apps, a basic health probe configuration that reports the app's availability to process requests (*liveness*) is sufficient to discover the status of the app.
+
+The basic configuration registers health check services and calls the Health Check Middleware to respond at a URL endpoint with a health response. By default, no specific health checks are registered to test any particular dependency or subsystem. The app is considered healthy if it's capable of responding at the health endpoint URL. The default response writer writes the status (`HealthCheckStatus`) as `text/plain` content back to the client. The response is a *200 Ok* status code, and the content is either healthy (`HealthCheckResult.Healthy`) or unhealthy (`HealthCheckResult.Unhealthy`).
+
+Register health check services with `AddHealthChecks` in `Startup.ConfigureServices`. Call Health Check Middleware in the app processing pipeline in `Startup.Configure`.
+
+In the sample app, the health check endpoint is created at `/health` (*BasicStartup.cs*):
+
+[!code-csharp[](health-checks/samples/2.x/HealthChecksSample/BasicStartup.cs?name=snippet1&highlight=5,10)]
+
+To run the basic configuration scenario using the sample app, execute the following command from the project's folder in a command shell:
+
+```console
+dotnet run --scenario basic
+```
+
+### Docker example
+
+[Docker](xref:host-and-deploy/docker/index) offers a built-in `HEALTHCHECK` directive that can be used to check the status of an app that uses the basic health check configuration:
+
+```
+HEALTHCHECK CMD curl --fail http://localhost:5000/health || exit
+```
+
+## Create health checks
+
+Health checks are created by implementing the `IHealthCheck` interface. The `IHealthCheck.CheckHealthAsync` method returns a `Task<HealthCheckResult>` of either `Healthy`, `Unhealthy`, `Degraded`, or `Failed`. The result is written as a plain text response with a configurable status code (configuration is described in the [Health check options](#health-check-options) section).
+
+The following `ExampleHealthCheck` class demonstrates the layout of a health check:
+
+```csharp
+public class ExampleHealthCheck : IHealthCheck
+{
+    public ExampleHealthCheck()
+    {
+        // Use DI to supply any required services to the health check.
+    }
+
+    public string Name => "example_check";
+
+    public Task<HealthCheckResult> CheckHealthAsync(
+        CancellationToken cancellationToken = default(CancellationToken))
+    {
+        // Execute health check logic here. This example sets a dummy
+        // variable to true.
+        var healthCheckResultHealthy = true;
+
+        // Return a HealthCheckResult value depending on the outcome of the
+        // check(s). Other HealthCheckResult outcomes not shown in this
+        // example are Degraded and Failed.
+        if (healthCheckResultHealthy)
+        {
+            return Task.FromResult(
+                HealthCheckResult.Healthy("The check indicates a healthy state."));
```

Comment:
This is a little out of date. We've pivoted on a few things since preview 3.

Here are the two important things to understand.
1. The application developer can configure the health check name and severity
2. The health check author can return pass/fail + a description + optional key value pairs


------

Here's a dump of some guidance I sent to the author of BeatPulse - some of it is for developing a **library** of health checks. Incorporate this as you see fit. 

> Some feedback on health check implementations:


> In general you don't need to log inside a health check. We do boilerplate logging for success and failure, so a log statement like "health check XYZ has failed" is redundant with that the developer gets by default. If you do have detailed things you want to report, that's what the description is for - which we also include in logs. I really struggle to think of a case where you'd need to log inside a health check implementation. Let me know if you have one, maybe I'm wrong 😂


> I would also suggest against put multiple checks inside a single implementation, such as ... where it loops through a list of URLs. If you just treat each URL as a separate check then will have already have a good logging and reporting experience. Let me know if you think there's a good case for this. It really seems to me like as an app developer I would want to see each URL that I'm checking pass/fail independently.

> Some feedback on AddXyzCheck methods:

> Your 'AddXyzCheck' methods accept the 'name' and 'failureSeverity' as parameters. It's of course cool and good to have default values for these as well - but it's intended that the application developer has control over these. This is important because the names need to be unique. So if you have multiple storage accounts, they can have their own check/logging/reporting/severity/tags. The same goes for tags, I'd suggest not providing any by default, and allow the developer to use as they desire.


> You can also use named options if it's a good fit for your scenario.  If you end up having any health checks that need DI and need access to data passed in by the caller its a good solution.


> You can see an example of all of this feedback via the EF health check I recently added to our repo: https://github.com/aspnet/Diagnostics/blob/release/2.2/src/Microsoft.Extensions.Diagnostics.HealthChecks.EntityFrameworkCore/DependencyInjection/EntityFrameworkCoreHealthChecksBuilderExtensions.cs

Reply:
> This is a little out of date.

I think you meant 'since preview2', right? Preview 3 isn't out yet AFAIK.

So, we should sit on this after all?? ... I should pull down Preview 3 when it lands and update this topic with the latest bits/approach.

Reply:
> no more planned changes for preview3

Cool beans ... I'll get preview3 when it lands and make the updates and ping you back when it's ready for another :eye:. We'll have time to wrap it up before RTM.